gh-138307: Update the Ellipsis documentation (#138306)
python/cpython: Doc/library/stdtypes.rst

The Ellipsis Object
-------------------

> This object is commonly used as a placeholder of another objects, values or even
> instructions.  It supports no

Reviewer: ```suggestion
This object is commonly used as a placeholder for other objects, values or even
instructions.  It supports no
```

special operations.  There is exactly one ellipsis object, named
:const:`Ellipsis` (a built-in name).  ``type(Ellipsis)()`` produces the
:const:`Ellipsis` singleton.

It is written as ``Ellipsis`` or ``...``.

> For instance, at the standard library and its documentation, ``Ellipsis`` can

Reviewer: ```suggestion
For instance, in the standard library and its documentation, ``Ellipsis`` can
```

> appears in :ref:`pretty printers <prettyprinter-objects>`,

Reviewer 2: This gets awkward: the use in pretty printing is three dots, but it's not an Ellipsis object.  It's a typographic convention for omitted content (which is why Python uses three dots for this).

Author: I'll fix it.

Do you think we should add this third type to the glossary?

Reviewer 3: Likewise recursive_repr uses `...` but it's not the `Ellipsis` form of `...`.

Author: What about doctest.ELLIPIS? Now, I'm thinking it is three dots, not the Ellipsis object.

> :const:`documentation tests <doctest.ELLIPSIS>`,

Reviewer 2: `doctest.ELLIPSIS` is not an Ellipsis object.  It's yet another instance of using three dots to mean "something is missing here."  In a doctest, `...` matches any text, like `.*` in a regex.  

BTW, another use of three dots is the secondary REPL prompt:
```
>>> (
...
```
We can list out these "faux Ellipses" (not in those words) as a way of helping people who end up here because they were trying to figure out what something means.

I would structure it something like this:

```
In typical use, ``...`` as an Ellipsis appears in a few different places:
- In type annotations for :ref:`Callable arguments <annotating-callables>` or `tuple elements <find-the-reference>`
- As the body of a function instead of a :ref:`pass statement <tut-pass>`
- In third-party libraries, such as Numpy's slicing and striding <https://numpy.org/doc/stable/user/basics.indexing.html#slicing-and-striding>`_

Python also uses three dots in ways that are not Ellipsis objects:
- Doctest's :const:`doctest.ELLIPSIS`, as a pattern for missing content
- As the interactive prompt for "please finish your statement"

Lastly, the Python documentation often uses three dots in conventional English usage to mean omitted content, even in code examples that also use them as an Ellipsis.
```

> :ref:`type annotations <annotating-callables>`,

Reviewer 4: The link goes to callables (`Callable[..., T]` meaning a gradual type covering callables with any sets of arguments), but this is not the only use in typing. It's also used for variable-length tuples (`tuple[int, ...]`), to represent an omitted default value in stubs or overloads (`def f(x: int = ...) -> None:`), and as an empty body for a function in stubs and a few other contexts (`def f(x: int) -> None: ...`).

Reviewer 2: I definitely don't have the right links, i wanted to get the language right first.

or instead of :ref:`pass statement <tut-pass>`.

.. seealso::

   `NumPy's slicing and striding <https://numpy.org/doc/stable/user/basics.indexing.html#slicing-and-striding>`_

>       A known-well Ellipsis use on third packages is the slicing on the Numpy.

Reviewer 5: ```suggestion
      A well-known Ellipsis use in third party packages is for slicing in Numpy.
```
The whole sentence is a bit odd.

The Ellipsis Object
-------------------

This object is commonly used used to indicate that something is omitted.
It supports no special operations.  There is exactly one ellipsis object, named
:const:`Ellipsis` (a built-in name).  ``type(Ellipsis)()`` produces the
:const:`Ellipsis` singleton.

It is written as ``Ellipsis`` or ``...``.

In typical use, ``...`` as the ``Ellipsis`` object appears in a few different
places, for instance:

- In type annotations, such as :ref:`callable arguments <annotating-callables>`
  or :ref:`tuple elements <annotating-tuples>`.

- As the body of a function instead of a :ref:`pass statement <tut-pass>`.

- In third-party libraries, such as `Numpy's slicing and striding
  <https://numpy.org/doc/stable/user/basics.indexing.html#slicing-and-striding>`_.

Python also uses three dots in ways that are not ``Ellipsis`` objects, for instance:

- Doctest's :const:`ELLIPSIS <doctest.ELLIPSIS>`, as a pattern for missing content.

> - The default Python prompt of the :term:`interactive` shell when entering the
>   code for an indented code block, meaning "please finish your statement".

Reviewer 2: ```suggestion
- The default Python prompt of the :term:`interactive` shell when partial input is incomplete.
```
